<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="UTF-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<meta name="generator" content="Asciidoctor 2.0.16">
<title>Practical differential privacy on pangenomes</title>
<link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Open+Sans:300,300italic,400,400italic,600,600italic%7CNoto+Serif:400,400italic,700,700italic%7CDroid+Sans+Mono:400,700">
<style>
/*! Asciidoctor default stylesheet | MIT License | https://asciidoctor.org */
/* Uncomment the following line when using as a custom stylesheet */
/* @import "https://fonts.googleapis.com/css?family=Open+Sans:300,300italic,400,400italic,600,600italic%7CNoto+Serif:400,400italic,700,700italic%7CDroid+Sans+Mono:400,700"; */
html{font-family:sans-serif;-webkit-text-size-adjust:100%}
a{background:none}
a:focus{outline:thin dotted}
a:active,a:hover{outline:0}
h1{font-size:2em;margin:.67em 0}
b,strong{font-weight:bold}
abbr{font-size:.9em}
abbr[title]{cursor:help;border-bottom:1px dotted #dddddf;text-decoration:none}
dfn{font-style:italic}
hr{height:0}
mark{background:#ff0;color:#000}
code,kbd,pre,samp{font-family:monospace;font-size:1em}
pre{white-space:pre-wrap}
q{quotes:"\201C" "\201D" "\2018" "\2019"}
small{font-size:80%}
sub,sup{font-size:75%;line-height:0;position:relative;vertical-align:baseline}
sup{top:-.5em}
sub{bottom:-.25em}
img{border:0}
svg:not(:root){overflow:hidden}
figure{margin:0}
audio,video{display:inline-block}
audio:not([controls]){display:none;height:0}
fieldset{border:1px solid silver;margin:0 2px;padding:.35em .625em .75em}
legend{border:0;padding:0}
button,input,select,textarea{font-family:inherit;font-size:100%;margin:0}
button,input{line-height:normal}
button,select{text-transform:none}
button,html input[type=button],input[type=reset],input[type=submit]{-webkit-appearance:button;cursor:pointer}
button[disabled],html input[disabled]{cursor:default}
input[type=checkbox],input[type=radio]{padding:0}
button::-moz-focus-inner,input::-moz-focus-inner{border:0;padding:0}
textarea{overflow:auto;vertical-align:top}
table{border-collapse:collapse;border-spacing:0}
*,::before,::after{box-sizing:border-box}
html,body{font-size:100%}
body{background:#fff;color:rgba(0,0,0,.8);padding:0;margin:0;font-family:"Noto Serif","DejaVu Serif",serif;line-height:1;position:relative;cursor:auto;-moz-tab-size:4;-o-tab-size:4;tab-size:4;word-wrap:anywhere;-moz-osx-font-smoothing:grayscale;-webkit-font-smoothing:antialiased}
a:hover{cursor:pointer}
img,object,embed{max-width:100%;height:auto}
object,embed{height:100%}
img{-ms-interpolation-mode:bicubic}
.left{float:left!important}
.right{float:right!important}
.text-left{text-align:left!important}
.text-right{text-align:right!important}
.text-center{text-align:center!important}
.text-justify{text-align:justify!important}
.hide{display:none}
img,object,svg{display:inline-block;vertical-align:middle}
textarea{height:auto;min-height:50px}
select{width:100%}
.subheader,.admonitionblock td.content>.title,.audioblock>.title,.exampleblock>.title,.imageblock>.title,.listingblock>.title,.literalblock>.title,.stemblock>.title,.openblock>.title,.paragraph>.title,.quoteblock>.title,table.tableblock>.title,.verseblock>.title,.videoblock>.title,.dlist>.title,.olist>.title,.ulist>.title,.qlist>.title,.hdlist>.title{line-height:1.45;color:#7a2518;font-weight:400;margin-top:0;margin-bottom:.25em}
div,dl,dt,dd,ul,ol,li,h1,h2,h3,#toctitle,.sidebarblock>.content>.title,h4,h5,h6,pre,form,p,blockquote,th,td{margin:0;padding:0}
a{color:#2156a5;text-decoration:underline;line-height:inherit}
a:hover,a:focus{color:#1d4b8f}
a img{border:0}
p{line-height:1.6;margin-bottom:1.25em;text-rendering:optimizeLegibility}
p aside{font-size:.875em;line-height:1.35;font-style:italic}
h1,h2,h3,#toctitle,.sidebarblock>.content>.title,h4,h5,h6{font-family:"Open Sans","DejaVu Sans",sans-serif;font-weight:300;font-style:normal;color:#ba3925;text-rendering:optimizeLegibility;margin-top:1em;margin-bottom:.5em;line-height:1.0125em}
h1 small,h2 small,h3 small,#toctitle small,.sidebarblock>.content>.title small,h4 small,h5 small,h6 small{font-size:60%;color:#e99b8f;line-height:0}
h1{font-size:2.125em}
h2{font-size:1.6875em}
h3,#toctitle,.sidebarblock>.content>.title{font-size:1.375em}
h4,h5{font-size:1.125em}
h6{font-size:1em}
hr{border:solid #dddddf;border-width:1px 0 0;clear:both;margin:1.25em 0 1.1875em}
em,i{font-style:italic;line-height:inherit}
strong,b{font-weight:bold;line-height:inherit}
small{font-size:60%;line-height:inherit}
code{font-family:"Droid Sans Mono","DejaVu Sans Mono",monospace;font-weight:400;color:rgba(0,0,0,.9)}
ul,ol,dl{line-height:1.6;margin-bottom:1.25em;list-style-position:outside;font-family:inherit}
ul,ol{margin-left:1.5em}
ul li ul,ul li ol{margin-left:1.25em;margin-bottom:0}
ul.square li ul,ul.circle li ul,ul.disc li ul{list-style:inherit}
ul.square{list-style-type:square}
ul.circle{list-style-type:circle}
ul.disc{list-style-type:disc}
ol li ul,ol li ol{margin-left:1.25em;margin-bottom:0}
dl dt{margin-bottom:.3125em;font-weight:bold}
dl dd{margin-bottom:1.25em}
blockquote{margin:0 0 1.25em;padding:.5625em 1.25em 0 1.1875em;border-left:1px solid #ddd}
blockquote,blockquote p{line-height:1.6;color:rgba(0,0,0,.85)}
@media screen and (min-width:768px){h1,h2,h3,#toctitle,.sidebarblock>.content>.title,h4,h5,h6{line-height:1.2}
h1{font-size:2.75em}
h2{font-size:2.3125em}
h3,#toctitle,.sidebarblock>.content>.title{font-size:1.6875em}
h4{font-size:1.4375em}}
table{background:#fff;margin-bottom:1.25em;border:1px solid #dedede;word-wrap:normal}
table thead,table tfoot{background:#f7f8f7}
table thead tr th,table thead tr td,table tfoot tr th,table tfoot tr td{padding:.5em .625em .625em;font-size:inherit;color:rgba(0,0,0,.8);text-align:left}
table tr th,table tr td{padding:.5625em .625em;font-size:inherit;color:rgba(0,0,0,.8)}
table tr.even,table tr.alt{background:#f8f8f7}
table thead tr th,table tfoot tr th,table tbody tr td,table tr td,table tfoot tr td{line-height:1.6}
h1,h2,h3,#toctitle,.sidebarblock>.content>.title,h4,h5,h6{line-height:1.2;word-spacing:-.05em}
h1 strong,h2 strong,h3 strong,#toctitle strong,.sidebarblock>.content>.title strong,h4 strong,h5 strong,h6 strong{font-weight:400}
.center{margin-left:auto;margin-right:auto}
.stretch{width:100%}
.clearfix::before,.clearfix::after,.float-group::before,.float-group::after{content:" ";display:table}
.clearfix::after,.float-group::after{clear:both}
:not(pre).nobreak{word-wrap:normal}
:not(pre).nowrap{white-space:nowrap}
:not(pre).pre-wrap{white-space:pre-wrap}
:not(pre):not([class^=L])>code{font-size:.9375em;font-style:normal!important;letter-spacing:0;padding:.1em .5ex;word-spacing:-.15em;background:#f7f7f8;border-radius:4px;line-height:1.45;text-rendering:optimizeSpeed}
pre{color:rgba(0,0,0,.9);font-family:"Droid Sans Mono","DejaVu Sans Mono",monospace;line-height:1.45;text-rendering:optimizeSpeed}
pre code,pre pre{color:inherit;font-size:inherit;line-height:inherit}
pre>code{display:block}
pre.nowrap,pre.nowrap pre{white-space:pre;word-wrap:normal}
em em{font-style:normal}
strong strong{font-weight:400}
.keyseq{color:rgba(51,51,51,.8)}
kbd{font-family:"Droid Sans Mono","DejaVu Sans Mono",monospace;display:inline-block;color:rgba(0,0,0,.8);font-size:.65em;line-height:1.45;background:#f7f7f7;border:1px solid #ccc;border-radius:3px;box-shadow:0 1px 0 rgba(0,0,0,.2),inset 0 0 0 .1em #fff;margin:0 .15em;padding:.2em .5em;vertical-align:middle;position:relative;top:-.1em;white-space:nowrap}
.keyseq kbd:first-child{margin-left:0}
.keyseq kbd:last-child{margin-right:0}
.menuseq,.menuref{color:#000}
.menuseq b:not(.caret),.menuref{font-weight:inherit}
.menuseq{word-spacing:-.02em}
.menuseq b.caret{font-size:1.25em;line-height:.8}
.menuseq i.caret{font-weight:bold;text-align:center;width:.45em}
b.button::before,b.button::after{position:relative;top:-1px;font-weight:400}
b.button::before{content:"[";padding:0 3px 0 2px}
b.button::after{content:"]";padding:0 2px 0 3px}
p a>code:hover{color:rgba(0,0,0,.9)}
#header,#content,#footnotes,#footer{width:100%;margin:0 auto;max-width:62.5em;*zoom:1;position:relative;padding-left:.9375em;padding-right:.9375em}
#header::before,#header::after,#content::before,#content::after,#footnotes::before,#footnotes::after,#footer::before,#footer::after{content:" ";display:table}
#header::after,#content::after,#footnotes::after,#footer::after{clear:both}
#content{margin-top:1.25em}
#content::before{content:none}
#header>h1:first-child{color:rgba(0,0,0,.85);margin-top:2.25rem;margin-bottom:0}
#header>h1:first-child+#toc{margin-top:8px;border-top:1px solid #dddddf}
#header>h1:only-child,body.toc2 #header>h1:nth-last-child(2){border-bottom:1px solid #dddddf;padding-bottom:8px}
#header .details{border-bottom:1px solid #dddddf;line-height:1.45;padding-top:.25em;padding-bottom:.25em;padding-left:.25em;color:rgba(0,0,0,.6);display:flex;flex-flow:row wrap}
#header .details span:first-child{margin-left:-.125em}
#header .details span.email a{color:rgba(0,0,0,.85)}
#header .details br{display:none}
#header .details br+span::before{content:"\00a0\2013\00a0"}
#header .details br+span.author::before{content:"\00a0\22c5\00a0";color:rgba(0,0,0,.85)}
#header .details br+span#revremark::before{content:"\00a0|\00a0"}
#header #revnumber{text-transform:capitalize}
#header #revnumber::after{content:"\00a0"}
#content>h1:first-child:not([class]){color:rgba(0,0,0,.85);border-bottom:1px solid #dddddf;padding-bottom:8px;margin-top:0;padding-top:1rem;margin-bottom:1.25rem}
#toc{border-bottom:1px solid #e7e7e9;padding-bottom:.5em}
#toc>ul{margin-left:.125em}
#toc ul.sectlevel0>li>a{font-style:italic}
#toc ul.sectlevel0 ul.sectlevel1{margin:.5em 0}
#toc ul{font-family:"Open Sans","DejaVu Sans",sans-serif;list-style-type:none}
#toc li{line-height:1.3334;margin-top:.3334em}
#toc a{text-decoration:none}
#toc a:active{text-decoration:underline}
#toctitle{color:#7a2518;font-size:1.2em}
@media screen and (min-width:768px){#toctitle{font-size:1.375em}
body.toc2{padding-left:15em;padding-right:0}
#toc.toc2{margin-top:0!important;background:#f8f8f7;position:fixed;width:15em;left:0;top:0;border-right:1px solid #e7e7e9;border-top-width:0!important;border-bottom-width:0!important;z-index:1000;padding:1.25em 1em;height:100%;overflow:auto}
#toc.toc2 #toctitle{margin-top:0;margin-bottom:.8rem;font-size:1.2em}
#toc.toc2>ul{font-size:.9em;margin-bottom:0}
#toc.toc2 ul ul{margin-left:0;padding-left:1em}
#toc.toc2 ul.sectlevel0 ul.sectlevel1{padding-left:0;margin-top:.5em;margin-bottom:.5em}
body.toc2.toc-right{padding-left:0;padding-right:15em}
body.toc2.toc-right #toc.toc2{border-right-width:0;border-left:1px solid #e7e7e9;left:auto;right:0}}
@media screen and (min-width:1280px){body.toc2{padding-left:20em;padding-right:0}
#toc.toc2{width:20em}
#toc.toc2 #toctitle{font-size:1.375em}
#toc.toc2>ul{font-size:.95em}
#toc.toc2 ul ul{padding-left:1.25em}
body.toc2.toc-right{padding-left:0;padding-right:20em}}
#content #toc{border:1px solid #e0e0dc;margin-bottom:1.25em;padding:1.25em;background:#f8f8f7;border-radius:4px}
#content #toc>:first-child{margin-top:0}
#content #toc>:last-child{margin-bottom:0}
#footer{max-width:none;background:rgba(0,0,0,.8);padding:1.25em}
#footer-text{color:hsla(0,0%,100%,.8);line-height:1.44}
#content{margin-bottom:.625em}
.sect1{padding-bottom:.625em}
@media screen and (min-width:768px){#content{margin-bottom:1.25em}
.sect1{padding-bottom:1.25em}}
.sect1:last-child{padding-bottom:0}
.sect1+.sect1{border-top:1px solid #e7e7e9}
#content h1>a.anchor,h2>a.anchor,h3>a.anchor,#toctitle>a.anchor,.sidebarblock>.content>.title>a.anchor,h4>a.anchor,h5>a.anchor,h6>a.anchor{position:absolute;z-index:1001;width:1.5ex;margin-left:-1.5ex;display:block;text-decoration:none!important;visibility:hidden;text-align:center;font-weight:400}
#content h1>a.anchor::before,h2>a.anchor::before,h3>a.anchor::before,#toctitle>a.anchor::before,.sidebarblock>.content>.title>a.anchor::before,h4>a.anchor::before,h5>a.anchor::before,h6>a.anchor::before{content:"\00A7";font-size:.85em;display:block;padding-top:.1em}
#content h1:hover>a.anchor,#content h1>a.anchor:hover,h2:hover>a.anchor,h2>a.anchor:hover,h3:hover>a.anchor,#toctitle:hover>a.anchor,.sidebarblock>.content>.title:hover>a.anchor,h3>a.anchor:hover,#toctitle>a.anchor:hover,.sidebarblock>.content>.title>a.anchor:hover,h4:hover>a.anchor,h4>a.anchor:hover,h5:hover>a.anchor,h5>a.anchor:hover,h6:hover>a.anchor,h6>a.anchor:hover{visibility:visible}
#content h1>a.link,h2>a.link,h3>a.link,#toctitle>a.link,.sidebarblock>.content>.title>a.link,h4>a.link,h5>a.link,h6>a.link{color:#ba3925;text-decoration:none}
#content h1>a.link:hover,h2>a.link:hover,h3>a.link:hover,#toctitle>a.link:hover,.sidebarblock>.content>.title>a.link:hover,h4>a.link:hover,h5>a.link:hover,h6>a.link:hover{color:#a53221}
details,.audioblock,.imageblock,.literalblock,.listingblock,.stemblock,.videoblock{margin-bottom:1.25em}
details{margin-left:1.25rem}
details>summary{cursor:pointer;display:block;position:relative;line-height:1.6;margin-bottom:.625rem;-webkit-tap-highlight-color:transparent}
details>summary::before{content:"";border:solid transparent;border-left:solid;border-width:.3em 0 .3em .5em;position:absolute;top:.5em;left:-1.25rem;transform:translateX(15%)}
details[open]>summary::before{border:solid transparent;border-top:solid;border-width:.5em .3em 0;transform:translateY(15%)}
details>summary::after{content:"";width:1.25rem;height:1em;position:absolute;top:.3em;left:-1.25rem}
.admonitionblock td.content>.title,.audioblock>.title,.exampleblock>.title,.imageblock>.title,.listingblock>.title,.literalblock>.title,.stemblock>.title,.openblock>.title,.paragraph>.title,.quoteblock>.title,table.tableblock>.title,.verseblock>.title,.videoblock>.title,.dlist>.title,.olist>.title,.ulist>.title,.qlist>.title,.hdlist>.title{text-rendering:optimizeLegibility;text-align:left;font-family:"Noto Serif","DejaVu Serif",serif;font-size:1rem;font-style:italic}
table.tableblock.fit-content>caption.title{white-space:nowrap;width:0}
.paragraph.lead>p,#preamble>.sectionbody>[class=paragraph]:first-of-type p{font-size:1.21875em;line-height:1.6;color:rgba(0,0,0,.85)}
.admonitionblock>table{border-collapse:separate;border:0;background:none;width:100%}
.admonitionblock>table td.icon{text-align:center;width:80px}
.admonitionblock>table td.icon img{max-width:none}
.admonitionblock>table td.icon .title{font-weight:bold;font-family:"Open Sans","DejaVu Sans",sans-serif;text-transform:uppercase}
.admonitionblock>table td.content{padding-left:1.125em;padding-right:1.25em;border-left:1px solid #dddddf;color:rgba(0,0,0,.6);word-wrap:anywhere}
.admonitionblock>table td.content>:last-child>:last-child{margin-bottom:0}
.exampleblock>.content{border:1px solid #e6e6e6;margin-bottom:1.25em;padding:1.25em;background:#fff;border-radius:4px}
.exampleblock>.content>:first-child{margin-top:0}
.exampleblock>.content>:last-child{margin-bottom:0}
.sidebarblock{border:1px solid #dbdbd6;margin-bottom:1.25em;padding:1.25em;background:#f3f3f2;border-radius:4px}
.sidebarblock>:first-child{margin-top:0}
.sidebarblock>:last-child{margin-bottom:0}
.sidebarblock>.content>.title{color:#7a2518;margin-top:0;text-align:center}
.exampleblock>.content>:last-child>:last-child,.exampleblock>.content .olist>ol>li:last-child>:last-child,.exampleblock>.content .ulist>ul>li:last-child>:last-child,.exampleblock>.content .qlist>ol>li:last-child>:last-child,.sidebarblock>.content>:last-child>:last-child,.sidebarblock>.content .olist>ol>li:last-child>:last-child,.sidebarblock>.content .ulist>ul>li:last-child>:last-child,.sidebarblock>.content .qlist>ol>li:last-child>:last-child{margin-bottom:0}
.literalblock pre,.listingblock>.content>pre{border-radius:4px;overflow-x:auto;padding:1em;font-size:.8125em}
@media screen and (min-width:768px){.literalblock pre,.listingblock>.content>pre{font-size:.90625em}}
@media screen and (min-width:1280px){.literalblock pre,.listingblock>.content>pre{font-size:1em}}
.literalblock pre,.listingblock>.content>pre:not(.highlight),.listingblock>.content>pre[class=highlight],.listingblock>.content>pre[class^="highlight "]{background:#f7f7f8}
.literalblock.output pre{color:#f7f7f8;background:rgba(0,0,0,.9)}
.listingblock>.content{position:relative}
.listingblock code[data-lang]::before{display:none;content:attr(data-lang);position:absolute;font-size:.75em;top:.425rem;right:.5rem;line-height:1;text-transform:uppercase;color:inherit;opacity:.5}
.listingblock:hover code[data-lang]::before{display:block}
.listingblock.terminal pre .command::before{content:attr(data-prompt);padding-right:.5em;color:inherit;opacity:.5}
.listingblock.terminal pre .command:not([data-prompt])::before{content:"$"}
.listingblock pre.highlightjs{padding:0}
.listingblock pre.highlightjs>code{padding:1em;border-radius:4px}
.listingblock pre.prettyprint{border-width:0}
.prettyprint{background:#f7f7f8}
pre.prettyprint .linenums{line-height:1.45;margin-left:2em}
pre.prettyprint li{background:none;list-style-type:inherit;padding-left:0}
pre.prettyprint li code[data-lang]::before{opacity:1}
pre.prettyprint li:not(:first-child) code[data-lang]::before{display:none}
table.linenotable{border-collapse:separate;border:0;margin-bottom:0;background:none}
table.linenotable td[class]{color:inherit;vertical-align:top;padding:0;line-height:inherit;white-space:normal}
table.linenotable td.code{padding-left:.75em}
table.linenotable td.linenos{border-right:1px solid;opacity:.35;padding-right:.5em}
pre.pygments .lineno{border-right:1px solid;opacity:.35;display:inline-block;margin-right:.75em}
pre.pygments .lineno::before{content:"";margin-right:-.125em}
.quoteblock{margin:0 1em 1.25em 1.5em;display:table}
.quoteblock:not(.excerpt)>.title{margin-left:-1.5em;margin-bottom:.75em}
.quoteblock blockquote,.quoteblock p{color:rgba(0,0,0,.85);font-size:1.15rem;line-height:1.75;word-spacing:.1em;letter-spacing:0;font-style:italic;text-align:justify}
.quoteblock blockquote{margin:0;padding:0;border:0}
.quoteblock blockquote::before{content:"\201c";float:left;font-size:2.75em;font-weight:bold;line-height:.6em;margin-left:-.6em;color:#7a2518;text-shadow:0 1px 2px rgba(0,0,0,.1)}
.quoteblock blockquote>.paragraph:last-child p{margin-bottom:0}
.quoteblock .attribution{margin-top:.75em;margin-right:.5ex;text-align:right}
.verseblock{margin:0 1em 1.25em}
.verseblock pre{font-family:"Open Sans","DejaVu Sans",sans-serif;font-size:1.15rem;color:rgba(0,0,0,.85);font-weight:300;text-rendering:optimizeLegibility}
.verseblock pre strong{font-weight:400}
.verseblock .attribution{margin-top:1.25rem;margin-left:.5ex}
.quoteblock .attribution,.verseblock .attribution{font-size:.9375em;line-height:1.45;font-style:italic}
.quoteblock .attribution br,.verseblock .attribution br{display:none}
.quoteblock .attribution cite,.verseblock .attribution cite{display:block;letter-spacing:-.025em;color:rgba(0,0,0,.6)}
.quoteblock.abstract blockquote::before,.quoteblock.excerpt blockquote::before,.quoteblock .quoteblock blockquote::before{display:none}
.quoteblock.abstract blockquote,.quoteblock.abstract p,.quoteblock.excerpt blockquote,.quoteblock.excerpt p,.quoteblock .quoteblock blockquote,.quoteblock .quoteblock p{line-height:1.6;word-spacing:0}
.quoteblock.abstract{margin:0 1em 1.25em;display:block}
.quoteblock.abstract>.title{margin:0 0 .375em;font-size:1.15em;text-align:center}
.quoteblock.excerpt>blockquote,.quoteblock .quoteblock{padding:0 0 .25em 1em;border-left:.25em solid #dddddf}
.quoteblock.excerpt,.quoteblock .quoteblock{margin-left:0}
.quoteblock.excerpt blockquote,.quoteblock.excerpt p,.quoteblock .quoteblock blockquote,.quoteblock .quoteblock p{color:inherit;font-size:1.0625rem}
.quoteblock.excerpt .attribution,.quoteblock .quoteblock .attribution{color:inherit;font-size:.85rem;text-align:left;margin-right:0}
p.tableblock:last-child{margin-bottom:0}
td.tableblock>.content{margin-bottom:1.25em;word-wrap:anywhere}
td.tableblock>.content>:last-child{margin-bottom:-1.25em}
table.tableblock,th.tableblock,td.tableblock{border:0 solid #dedede}
table.grid-all>*>tr>*{border-width:1px}
table.grid-cols>*>tr>*{border-width:0 1px}
table.grid-rows>*>tr>*{border-width:1px 0}
table.frame-all{border-width:1px}
table.frame-ends{border-width:1px 0}
table.frame-sides{border-width:0 1px}
table.frame-none>colgroup+*>:first-child>*,table.frame-sides>colgroup+*>:first-child>*{border-top-width:0}
table.frame-none>:last-child>:last-child>*,table.frame-sides>:last-child>:last-child>*{border-bottom-width:0}
table.frame-none>*>tr>:first-child,table.frame-ends>*>tr>:first-child{border-left-width:0}
table.frame-none>*>tr>:last-child,table.frame-ends>*>tr>:last-child{border-right-width:0}
table.stripes-all tr,table.stripes-odd tr:nth-of-type(odd),table.stripes-even tr:nth-of-type(even),table.stripes-hover tr:hover{background:#f8f8f7}
th.halign-left,td.halign-left{text-align:left}
th.halign-right,td.halign-right{text-align:right}
th.halign-center,td.halign-center{text-align:center}
th.valign-top,td.valign-top{vertical-align:top}
th.valign-bottom,td.valign-bottom{vertical-align:bottom}
th.valign-middle,td.valign-middle{vertical-align:middle}
table thead th,table tfoot th{font-weight:bold}
tbody tr th{background:#f7f8f7}
tbody tr th,tbody tr th p,tfoot tr th,tfoot tr th p{color:rgba(0,0,0,.8);font-weight:bold}
p.tableblock>code:only-child{background:none;padding:0}
p.tableblock{font-size:1em}
ol{margin-left:1.75em}
ul li ol{margin-left:1.5em}
dl dd{margin-left:1.125em}
dl dd:last-child,dl dd:last-child>:last-child{margin-bottom:0}
ol>li p,ul>li p,ul dd,ol dd,.olist .olist,.ulist .ulist,.ulist .olist,.olist .ulist{margin-bottom:.625em}
ul.checklist,ul.none,ol.none,ul.no-bullet,ol.no-bullet,ol.unnumbered,ul.unstyled,ol.unstyled{list-style-type:none}
ul.no-bullet,ol.no-bullet,ol.unnumbered{margin-left:.625em}
ul.unstyled,ol.unstyled{margin-left:0}
ul.checklist>li>p:first-child{margin-left:-1em}
ul.checklist>li>p:first-child>.fa-square-o:first-child,ul.checklist>li>p:first-child>.fa-check-square-o:first-child{width:1.25em;font-size:.8em;position:relative;bottom:.125em}
ul.checklist>li>p:first-child>input[type=checkbox]:first-child{margin-right:.25em}
ul.inline{display:flex;flex-flow:row wrap;list-style:none;margin:0 0 .625em -1.25em}
ul.inline>li{margin-left:1.25em}
.unstyled dl dt{font-weight:400;font-style:normal}
ol.arabic{list-style-type:decimal}
ol.decimal{list-style-type:decimal-leading-zero}
ol.loweralpha{list-style-type:lower-alpha}
ol.upperalpha{list-style-type:upper-alpha}
ol.lowerroman{list-style-type:lower-roman}
ol.upperroman{list-style-type:upper-roman}
ol.lowergreek{list-style-type:lower-greek}
.hdlist>table,.colist>table{border:0;background:none}
.hdlist>table>tbody>tr,.colist>table>tbody>tr{background:none}
td.hdlist1,td.hdlist2{vertical-align:top;padding:0 .625em}
td.hdlist1{font-weight:bold;padding-bottom:1.25em}
td.hdlist2{word-wrap:anywhere}
.literalblock+.colist,.listingblock+.colist{margin-top:-.5em}
.colist td:not([class]):first-child{padding:.4em .75em 0;line-height:1;vertical-align:top}
.colist td:not([class]):first-child img{max-width:none}
.colist td:not([class]):last-child{padding:.25em 0}
.thumb,.th{line-height:0;display:inline-block;border:4px solid #fff;box-shadow:0 0 0 1px #ddd}
.imageblock.left{margin:.25em .625em 1.25em 0}
.imageblock.right{margin:.25em 0 1.25em .625em}
.imageblock>.title{margin-bottom:0}
.imageblock.thumb,.imageblock.th{border-width:6px}
.imageblock.thumb>.title,.imageblock.th>.title{padding:0 .125em}
.image.left,.image.right{margin-top:.25em;margin-bottom:.25em;display:inline-block;line-height:0}
.image.left{margin-right:.625em}
.image.right{margin-left:.625em}
a.image{text-decoration:none;display:inline-block}
a.image object{pointer-events:none}
sup.footnote,sup.footnoteref{font-size:.875em;position:static;vertical-align:super}
sup.footnote a,sup.footnoteref a{text-decoration:none}
sup.footnote a:active,sup.footnoteref a:active{text-decoration:underline}
#footnotes{padding-top:.75em;padding-bottom:.75em;margin-bottom:.625em}
#footnotes hr{width:20%;min-width:6.25em;margin:-.25em 0 .75em;border-width:1px 0 0}
#footnotes .footnote{padding:0 .375em 0 .225em;line-height:1.3334;font-size:.875em;margin-left:1.2em;margin-bottom:.2em}
#footnotes .footnote a:first-of-type{font-weight:bold;text-decoration:none;margin-left:-1.05em}
#footnotes .footnote:last-of-type{margin-bottom:0}
#content #footnotes{margin-top:-.625em;margin-bottom:0;padding:.75em 0}
.gist .file-data>table{border:0;background:#fff;width:100%;margin-bottom:0}
.gist .file-data>table td.line-data{width:99%}
div.unbreakable{page-break-inside:avoid}
.big{font-size:larger}
.small{font-size:smaller}
.underline{text-decoration:underline}
.overline{text-decoration:overline}
.line-through{text-decoration:line-through}
.aqua{color:#00bfbf}
.aqua-background{background:#00fafa}
.black{color:#000}
.black-background{background:#000}
.blue{color:#0000bf}
.blue-background{background:#0000fa}
.fuchsia{color:#bf00bf}
.fuchsia-background{background:#fa00fa}
.gray{color:#606060}
.gray-background{background:#7d7d7d}
.green{color:#006000}
.green-background{background:#007d00}
.lime{color:#00bf00}
.lime-background{background:#00fa00}
.maroon{color:#600000}
.maroon-background{background:#7d0000}
.navy{color:#000060}
.navy-background{background:#00007d}
.olive{color:#606000}
.olive-background{background:#7d7d00}
.purple{color:#600060}
.purple-background{background:#7d007d}
.red{color:#bf0000}
.red-background{background:#fa0000}
.silver{color:#909090}
.silver-background{background:#bcbcbc}
.teal{color:#006060}
.teal-background{background:#007d7d}
.white{color:#bfbfbf}
.white-background{background:#fafafa}
.yellow{color:#bfbf00}
.yellow-background{background:#fafa00}
span.icon>.fa{cursor:default}
a span.icon>.fa{cursor:inherit}
.admonitionblock td.icon [class^="fa icon-"]{font-size:2.5em;text-shadow:1px 1px 2px rgba(0,0,0,.5);cursor:default}
.admonitionblock td.icon .icon-note::before{content:"\f05a";color:#19407c}
.admonitionblock td.icon .icon-tip::before{content:"\f0eb";text-shadow:1px 1px 2px rgba(155,155,0,.8);color:#111}
.admonitionblock td.icon .icon-warning::before{content:"\f071";color:#bf6900}
.admonitionblock td.icon .icon-caution::before{content:"\f06d";color:#bf3400}
.admonitionblock td.icon .icon-important::before{content:"\f06a";color:#bf0000}
.conum[data-value]{display:inline-block;color:#fff!important;background:rgba(0,0,0,.8);border-radius:50%;text-align:center;font-size:.75em;width:1.67em;height:1.67em;line-height:1.67em;font-family:"Open Sans","DejaVu Sans",sans-serif;font-style:normal;font-weight:bold}
.conum[data-value] *{color:#fff!important}
.conum[data-value]+b{display:none}
.conum[data-value]::after{content:attr(data-value)}
pre .conum[data-value]{position:relative;top:-.125em}
b.conum *{color:inherit!important}
.conum:not([data-value]):empty{display:none}
dt,th.tableblock,td.content,div.footnote{text-rendering:optimizeLegibility}
h1,h2,p,td.content,span.alt,summary{letter-spacing:-.01em}
p strong,td.content strong,div.footnote strong{letter-spacing:-.005em}
p,blockquote,dt,td.content,span.alt,summary{font-size:1.0625rem}
p{margin-bottom:1.25rem}
.sidebarblock p,.sidebarblock dt,.sidebarblock td.content,p.tableblock{font-size:1em}
.exampleblock>.content{background:#fffef7;border-color:#e0e0dc;box-shadow:0 1px 4px #e0e0dc}
.print-only{display:none!important}
@page{margin:1.25cm .75cm}
@media print{*{box-shadow:none!important;text-shadow:none!important}
html{font-size:80%}
a{color:inherit!important;text-decoration:underline!important}
a.bare,a[href^="#"],a[href^="mailto:"]{text-decoration:none!important}
a[href^="http:"]:not(.bare)::after,a[href^="https:"]:not(.bare)::after{content:"(" attr(href) ")";display:inline-block;font-size:.875em;padding-left:.25em}
abbr[title]{border-bottom:1px dotted}
abbr[title]::after{content:" (" attr(title) ")"}
pre,blockquote,tr,img,object,svg{page-break-inside:avoid}
thead{display:table-header-group}
svg{max-width:100%}
p,blockquote,dt,td.content{font-size:1em;orphans:3;widows:3}
h2,h3,#toctitle,.sidebarblock>.content>.title{page-break-after:avoid}
#header,#content,#footnotes,#footer{max-width:none}
#toc,.sidebarblock,.exampleblock>.content{background:none!important}
#toc{border-bottom:1px solid #dddddf!important;padding-bottom:0!important}
body.book #header{text-align:center}
body.book #header>h1:first-child{border:0!important;margin:2.5em 0 1em}
body.book #header .details{border:0!important;display:block;padding:0!important}
body.book #header .details span:first-child{margin-left:0!important}
body.book #header .details br{display:block}
body.book #header .details br+span::before{content:none!important}
body.book #toc{border:0!important;text-align:left!important;padding:0!important;margin:0!important}
body.book #toc,body.book #preamble,body.book h1.sect0,body.book .sect1>h2{page-break-before:always}
.listingblock code[data-lang]::before{display:block}
#footer{padding:0 .9375em}
.hide-on-print{display:none!important}
.print-only{display:block!important}
.hide-for-print{display:none!important}
.show-for-print{display:inherit!important}}
@media amzn-kf8,print{#header>h1:first-child{margin-top:1.25rem}
.sect1{padding:0!important}
.sect1+.sect1{border:0}
#footer{background:none}
#footer-text{color:rgba(0,0,0,.6);font-size:.9em}}
@media amzn-kf8{#header,#content,#footnotes,#footer{padding:0}}
</style>
</head>
<body class="article">
<div id="header">
<h1>Practical differential privacy on pangenomes</h1>
</div>
<div id="content">
<div id="preamble">
<div class="sectionbody">
<div class="paragraph">
<p>Erik Garrison  &lt;<a href="mailto:erik.garrison@gmail.com">erik.garrison@gmail.com</a>&gt;</p>
</div>
<div class="paragraph">
<p><em>tl;dr</em>: We implement a differentially private haplotype sampling method in a <a href="https://github.com/pangenome/odgi">pangenomics toolkit</a>.
It projects <a href="https://en.wikipedia.org/wiki/Differential_privacy#%CE%B5-differential_privacy">\(\epsilon\)-differentially private</a> synthetic pangenome <a href="https://doi.org/10.1038/nbt.4227">variation graphs</a> out of pangenomes built from complete haplotype-resolved assemblies like those made in the <a href="https://humanpangenome.org/">Human Pangenome project</a>.
We generate \(\epsilon\)-differentially private graphs from the human <a href="https://en.wikipedia.org/wiki/Major_histocompatibility_complex">major_histocompatibility_complex (MHC)</a>, and use these to explore the effects of algorithm paramaters on output.</p>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_building_a_globally_inclusive_pangenome">Building a globally inclusive pangenome</h2>
<div class="sectionbody">
<div class="paragraph">
<p>Our goal the privvg project is to enable researchers to share sequences from (pan)genomic cohorts without violating individual privacy guarantees.
The release of this information can make it easier to understand the sequence of a new genome which we sequence or analyze.
And, it can give us a foundational reference describing variation and evolution in humans.</p>
</div>
<div class="paragraph">
<p>In establishing a method to share sequences from private cohorts publicly, we hope to spark the development of a universal human pangenome that can be built fully in the open using material that represents the group properties of cohorts which are otherwise impossible to access.</p>
</div>
<div class="sect2">
<h3 id="_sharing_genomic_information_has_major_social_benefits">Sharing genomic information has major social benefits</h3>
<div class="paragraph">
<p>When we observe a new genome, we do so with DNA sequence reads that are much smaller than the full genome.
To place them in the context of biological features, we compare them to a complete set of genomes.
Our sequencing and analysis costs can be reduced dramatically if we compare the sequence reads to a reference genome, or pangenome (many genomes).
A pangenome representing many individuals can show us the frequency of a given <a href="https://en.wikipedia.org/wiki/Allele">allele</a>, which is critical in the analysis of rare genomic disease cases.</p>
</div>
<div class="paragraph">
<p>There are significant practical benefits to data sharing.
But, the standard in human genomics is to <em>not</em> share information, as the release of genomic information can pose a major privacy risk.
Some sharing does occur, for instance, many medical cohorts are able to <a href="https://gnomad.broadinstitute.org/">release summary statistics</a> about <a href="https://en.wikipedia.org/wiki/Single-nucleotide_polymorphism">SNPs</a>, such as their frequency.
Unfortunately, this is very limited information.
We would rather know frequencies of the alleles of whole genes—the functional units of the genome—which may be 10-100Kbp.
Furthermore, given that the cost to build a complete <em>de novo</em> assembly of a human genome is rapidly dropping <a href="#nurk2022">[nurk2022]</a>, we see a need to develop a method that lets the privacy-preserving data release that provides access to the full sequence space of the pangenome, not just point mutations relative to a reference genome.</p>
</div>
</div>
<div class="sect2">
<h3 id="_enabling_privacy_preserving_sharing_with_differential_privacy">Enabling privacy-preserving sharing with differential privacy</h3>
<div class="paragraph">
<p>Large medical cohorts with associations between genomes and phenotypes usually only provide controlled data access to trusted researchers.
Our goal is to establish a standard whereby a they could release a fully-public transformation of this controlled data for global use by anyone.
This would thus provide global biomedical utility without significant risk to study participants.
We do so by applying tools from <a href="https://en.wikipedia.org/wiki/Differential_privacy">differential privacy</a>.</p>
</div>
<div class="paragraph">
<p>Differential privacy is an approach to data release that which allows for the description of group characteristics without revealing information about single individuals.
It quantifies privacy loss caused by the release of information, allowing us to reason about the risks that a particular data sharing model poses to individual privacy.
We build on decades of work on differential privacy, which has yielded well-defined models of differentially private data publication <a href="#dwork2014">[dwork2014]</a>.
Our work is similar to approaches that have been used for trajectory data release, but differs in the unique biomedical context and properties of the graph data structures we use.
This has led to the need for an application-specific implementation.</p>
</div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_synthetic_differentially_private_pangenomes">Synthetic differentially private pangenomes</h2>
<div class="sectionbody">
<div class="paragraph">
<p>Differential privacy provides tools that allow us to generate privacy-preserving synthetic databases out of real ones.
We pursue this approach because we judge that it is easier to share static databases than organize access to differentially private queries.
If we provide a tool to produce a differentially private synthetic pangenome, a genomics research project could release a pangenome that can be reused indefinitely by external researchers.
Reuse of this database would not constitute increased privacy risk to individuals <a href="#dwork2014">[dwork2014]</a>.</p>
</div>
<div class="sect2">
<h3 id="_applying_the_exponential_mechanism">Applying the exponential mechanism</h3>
<div class="paragraph">
<p>We say that a mechanism to draw synthetic database \(x\) from range \(\mathcal R\) is \(\epsilon\)-differentially private.
In the abstract, this means that, if we had another <em>neighboring</em> synthetic database \(y\) differing from \(x\) in the exclusion of a given individual, we would expect them to differ by less than \(\epsilon\).
These approaches work by adding controlled, shaped noise to the elements and contents of queries or synthetic databases generated out of private data.</p>
</div>
<div class="paragraph">
<p>To develop such a synthetic database, we have a unique situation.
In contrast to tabular data, we have a sequence collection, and the sequences have been compressed into a graph <a href="#garrison2022">[garrison2022]</a>.
Genomes are walks through this graph.
Intuitively, a differentially private query of a pangenome could expose information about sequences while adding noise to the walk.
However, such noise could easily destroy the value of the released data, generating nonsense haplotypes that may not be viable or have a meaningful frequency in the population.</p>
</div>
<div class="paragraph">
<p>The <em>exponential mechanism</em> was designed for situations like this, where we want to choose the best response for a query but adding noise directly to the computed result can utterly destroy its value.
Following this approach, we&#8217;ve designed the <a href="https://privvg.github.io/2022/06/13/Differential-Privacy.htmll">privvg algorithm</a>.
In essence, this approach samples haplotypes from a pangenome graph (formally, a <em>variation graph</em>) by following a walk guided by the exponential mechanism.</p>
</div>
</div>
<div class="sect2">
<h3 id="_the_shape_of_the_privacy_preserving_pangenome_graph">The shape of the privacy-preserving pangenome graph</h3>
<div class="paragraph">
<p>The exponential mechanism operates on the exponentiated <em>utility</em> of query results, releasing randomly sampled data elements \(\propto \exp(u)\).
Exponentiation helps to provide differential privacy guarantees, while utility is a social construct that lets us define which database elements are more important.</p>
</div>
<div class="paragraph">
<p>In our context, we assume that the utility of a haplotype observation (equivalently: allele, or genomic subsequence) is the log-frequency of this sequence in the population.
Intuitively, more common haplotypes are more likely to be found in other individuals.
The better that these are represented in our pangenome, the better the pangenome will contain or represent a new genome.
This makes them more useful.
However, the incremental benefit of one high frequency allele over another is proportionally less than for lower-frequency alleles (thus, we take the logarithm of frequency).
We also would like to not expose too many rare haplotypes, as they may allow for privacy-destructive attacks, so their release would decrease utility.</p>
</div>
<div class="paragraph">
<p>By guiding the exponential mechanism using <em>haplotype frequency</em> as utility, we will obtain a pangenome that represents group characteristics and common sequences in the pangenome.
Bear in mind that "common" here might mean "shared by a few individuals", and could in fact be very rare in a fractional sense when we apply privvg to a very large cohort pangenome.</p>
</div>
</div>
<div class="sect2">
<h3 id="_odgi_a_command_line_analysis_toolkit_for_pangenomics">odgi: a command line analysis toolkit for pangenomics</h3>
<div class="paragraph">
<p>The <strong>optimized dynamic graph interface</strong>, <a href="https://github.com/pangenome/odgi"><code>odgi</code></a> <a href="#guarracino2022">[guarracino2022]</a>, provides a command line interface to diverse operations on pangenome graphs <a href="#eizenga2020">[eizenga2020]</a>.
It allows for transformation and dynamic construction of new graphs based on others and data associated with their embedded genome paths.
It&#8217;s designed in the style of <a href="https://bedtools.readthedocs.io/en/latest/index.html"><code>bedtools</code></a>—a well-known bioinformatic toolkit for working with positions, ranges, and alignments on genomes.
But, instead working on genomes, <code>odgi</code> works on <em>pangenomes</em>, and specifically graphs that represent them.</p>
</div>
<div class="paragraph">
<p>We developed this toolkit to support work in the <a href="https://humanpangenome.org/">Human Pangenome project</a>, where it has provided an essential interface to the graphs we built from 44 human genomes (some 90 human haplotypes when including the two reference genomes) <a href="#liao2022">[liao2022]</a>.
Our previous explorations have focused on the <a href="https://github.com/jltsiren/gbwt">GBWT</a>, but in the interest of avoiding additional software dependencies and indexing steps, we decided to directly implement the method no top of the dynamic graph data structure offered by <code>odgi</code>.</p>
</div>
</div>
<div class="sect2">
<h3 id="_our_plan_of_attack">Our plan of attack</h3>
<div class="paragraph">
<p><code>odgi</code> provides an ideal delivery vehicle for our research into differentially private pangenome transformations.
Previously, we developed  <a href="https://privvg.github.io/2022/06/13/Differential-Privacy.htmll">a prototype</a> based on the graph Burrows-Wheeler transform (GBWT).
To simplify development and application of our approach, we opted to build directly on top of the dynamic graph model in <code>odgi</code>.</p>
</div>
<div class="sect3">
<h4 id="_parameters">Parameters</h4>
<div class="paragraph">
<p>We take a few parameters:</p>
</div>
<div class="dlist unordered">
<dl>
<dt><code>e</code> (epsilon)</dt>
<dd>
<p>Differential privacy parameter \(\epsilon\), expressed in terms of utility, which is allele observation counts in haplotypes.</p>
</dd>
<dt><code>d</code> (target-depth)</dt>
<dd>
<p>A target depth to which we cover the graph. We continue sampling haplotypes until we collect enough sequence to achieve this coverage. A future enhancement will be to use a privacy budget to determine the total depth of sampling, rather than a target depth. In practice, we&#8217;ll probably want this depth to be lower than that of our source data.</p>
</dd>
<dt><code>c</code> (min-hap-freq)</dt>
<dd>
<p>Released haplotypes must have at least this frequency in our base grap. At least 2x is required, or we run into singularities in the exponential mechanism, where the change in utility \(\Delta u\) becomes infinite.</p>
</dd>
<dt><code>b</code> (bp-target)</dt>
<dd>
<p>A minimum length of a sampled haplotype. The sampling algorithm runs until we reach this length, at which point we emit the sampled haplotype so long as it&#8217;s at or above our minimum haplotype frequency.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_sampling_algorithm">Sampling algorithm</h4>
<div class="paragraph">
<p>Our sampling algorithm follows a basic approach.
(1) We randomly select a node, and collect all the path steps across it.
(2) We bundle these paths depending on where they next go, and select a bundle using the exponential mechanism.
(3) The process iterates, using the bundle of open paths, until we reach a length or frequency based stopping condition.
(4) If our haplotype frequency is too low, we do not emit the haplotype.
(5) If it&#8217;s high enough and our length is sufficient, we emit the haplotype and count its length towards our target.</p>
</div>
<div class="paragraph">
<p>For greater clarity, we provide it in pseudocode:</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="highlight"><code>target_length := d * graph_length
sampled_length := 0
while sampled_length &lt; target_length:
  handle := get a random node and orientation
  ranges := {}
  # initial state
  for each step on handle:
    ranges.insert((step, step))
  walk_length := 0
  while ranges is not empty:
    # determine potential next steps
    nexts := map node -&gt; {}
    for range in ranges:
      next_step := get_next_step(range.end)
      next_handle := get_handle_of_step(next_step)
      nexts[next_handle].insert((range.begin, next_step))
    # compute exponential mechanism weights for extension
    sum_weights := 0
    weights := []
    for next in nexts:
      # log-frequency of the extension is utility
      utility := log1p(next.size())
      # sensitivity, or the effect of removing an individual on utility
      delta_utility := utility - log1p(next.size()-1)
      # weight for sampling
      weight := exp((epsilon * utility) / (2 * delta_utility))
      weights.append((weight, next))
      sum_weights += weight
    # use weighted random sampling to compute the next extension
    sample := random_0_1() * sum_weights
    x := 0
    opt := null
    for weight in weights:
      if x + w.first &gt;= sample:
        opt := w.second
      x += weight.first
    # now opt is our next step, so we recurse onto it
    ranges = nexts[opt]
    # update our sampled length
    walk_length += get_length(opt)
    # and determine if we've reached a stopping point
    if ranges.size() &lt; min_hap_freq:
      # our frequency is too low, violating our parameters
      break
    # our stopping condition
    if ranges.size() &gt;= min_hap_freq and walk_length &gt;= bp_target:
      # update our sampled length
      sampled_length += walk_length
      # to avoid orientation bias, we emit a randomly-sampled representative
      emit_sampled_haplotype(ranges.random())</code></pre>
</div>
</div>
</div>
</div>
<div class="sect2">
<h3 id="_the_privvg_algorithm_lands_in_odgi">The privvg algorithm lands in odgi!</h3>
<div class="paragraph">
<p>We&#8217;ve <a href="https://github.com/pangenome/odgi/blob/master/src/algorithms/diffpriv.cpp">implemented this algorithm in odgi</a>!
To ensure scalability, we run the sampling algorithm in parallel over the graph, with each thread accumulating samples independently.
(This is currently fully random, but a future improvement will allow for deterministic sampling, which is critical for testing.)
Tests on the human MHC indicate runtimes that are more than acceptable for the application to the entire draft Human Pangenome (HPRCy1) <a href="#liao2022">[liao2022]</a>.
The rest of this post will use this practical tool to generate synthetic differentially private pangenomes and explore their features relative to key parameters.</p>
</div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_exploring_privvg_parameters_with_the_human_mhc">Exploring privvg parameters with the human MHC</h2>
<div class="sectionbody">
<div class="paragraph">
<p>To get a feeling for how the various parameters affect the shape of emitted graphs, we can work through a few tests based on one of the more interesting and diverse regions of the human pangenome: the <a href="https://en.wikipedia.org/wiki/Major_histocompatibility_complex">major_histocompatibility_complex</a> or MHC.
The idea here is to get a taste for what these key parameters can do to the synthetic differentially private haplotype set that we produce.
We&#8217;ll use tools from <code>odgi</code> to get access to the graph itself.</p>
</div>
<div class="sect2">
<h3 id="_looking_at_mhc_pangenome_graph">Looking at MHC pangenome graph</h3>
<div class="paragraph">
<p>First, we built a graph of the MHC using the <a href="https://github.com/pangenome/pggb">PanGenome Graph Builder, <code>pggb</code></a>.
(We collected MHC sequences <a href="https://github.com/pangenome/MHC">as described in this short tutorial</a>, and then built the graph using <code>pggb -n 90 -p 90 -s 50k -k 29</code> with 0.3.1+237bf05.)</p>
</div>
<div class="paragraph">
<p>We download, unpack it, and build an <code>odgi</code> graph from it.</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="highlight"><code class="language-shell" data-lang="shell">wget https://f004.backblazeb2.com/file/pangenome/HPRCy1/HPRCy1v2.MHC.fa.ce6f12f.e9aeea8.0ead406.smooth.final.gfa.zst
zstdcat HPRCy1v2.MHC.fa.ce6f12f.e9aeea8.0ead406.smooth.final.gfa.zst &gt;mhc.gfa
odgi build -g mhc.gfa -o mhc.og -t 16</code></pre>
</div>
</div>
<div class="paragraph">
<p>Let&#8217;s check some basic statistics:</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="highlight"><code class="language-shell" data-lang="shell">odgi stats -i mhc.og -S
#length nodes   edges   paths   steps
5372615 257163  359123  126     14873189</code></pre>
</div>
</div>
<div class="paragraph">
<p>The graph is 5.37Mbp long, with 126 paths (there are 90 haplotypes, but some haplotypes have multiple contigs, each corresponding to a separate path), and 14M steps.</p>
</div>
<div class="paragraph">
<p>For diagnostics, wec an get a 1D visualization of the graph with <code>odgi viz</code>.
This shows us a kind of binary matrix.
Across the x-axis we have nodes in the graph.
Across the y-axis we have paths through the graph.
Each path is a specific contig, representing a part of a haplotype of a human genome.
Where we have white, the path is not crossing the given region of the graph, but where a color (based on a hash of the path name) is assigned, we can see that the path does occur.
In effect, this shows that many haplotypes cover the entire MHC.</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="highlight"><code class="language-shell" data-lang="shell">odgi viz -i mhc.og -o mhc.png</code></pre>
</div>
</div>
<div class="imageblock">
<div class="content">
<img src="../../../assets/mhc_odgi_viz.png" alt="odgi viz of MHC">
</div>
</div>
<div class="paragraph">
<p>The complex region to the right side of the visualization corresponds to the MHC class II genes.</p>
</div>
<div class="paragraph">
<p>We can visualize this too, with <a href="https://github.com/chfi/gfaestus">gfaestus</a>.
The orientation and order are very similar to the above 1D visualization, letting us see how the graph looks in 2D.</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="highlight"><code class="language-shell" data-lang="shell">odgi layout -i mhc.og -T mhc.og.lay.tsv -t 16 -P
gfaestus mhc.gfa mhc.og.lay.tsv</code></pre>
</div>
</div>
<div class="imageblock">
<div class="content">
<img src="../../../assets/mhc_gfaestus_full.png" alt="gfaestus of full MHC">
</div>
</div>
<div class="paragraph">
<p>Here, the splines we see correspond to sequences in the graph.
Each node has a specific color, leading to the rainbow patterns that we see.
Each change in color thus indicates a new allele in the pangenome.</p>
</div>
<div class="paragraph">
<p>Zooming in on the MHC class II genes shows a very complex structure.
This is one of the most highly-diverged regions in the human genome.
<a href="https://en.wikipedia.org/wiki/Balancing_selection">Balancing selection</a> drives diversity here, as the genes in the MHC encode for features of the immune system.
It is very advantageous for the human population to not have the same alleles here, so that our response to pathogens varies and it is not easy for a single pathogen to cause illness in all people.</p>
</div>
<div class="imageblock">
<div class="content">
<img src="../../../assets/mhc_gfaestus_mhc_d.png" alt="gfaestus of MHC class II">
</div>
</div>
<div class="paragraph">
<p>Finally, we can get a compressed mode output from <code>odgi viz</code> which shows us a kind of heatmap across the graph space.
For the full MHC graph, this is very boring: we have coverage almost everywhere:</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="highlight"><code class="language-shell" data-lang="shell">odgi viz -i mhc.og -o mhc.compr.png -O</code></pre>
</div>
</div>
<div class="imageblock">
<div class="content">
<img src="../../../assets/mhc_odgi_viz.compr.png" alt="odgi viz compressed MHC">
</div>
</div>
<div class="paragraph">
<p>We&#8217;ll use this when making privvg transformations of the graph to get a sense of what we lose with different parameter combinations.</p>
</div>
</div>
<div class="sect2">
<h3 id="_applying_privvg_to_the_mhc_pangenome">Applying privvg to the MHC pangenome</h3>
<div class="paragraph">
<p><code>odgi priv</code> provides the primary interface to apply the privvg algorithm to a graph.</p>
</div>
<div class="paragraph">
<p>Here we extract a haplotype sample of 10kbp sequences, targeting 100x total coverage, and emitting a progress message:</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="highlight"><code class="language-shell" data-lang="shell">odgi priv -i mhc.og -o mhc_priv.og -t 16 -b 10000 -d 100 -e 0.01 -P
odgi viz -i mhc_priv.og -o x.png -O
odgi stats -i mhc_priv.og -S</code></pre>
</div>
</div>
<div class="listingblock">
<div class="content">
<pre class="highlight"><code class="language-shell" data-lang="shell">#length nodes   edges   paths   steps
5372615 257163  292605  53012   19374084</code></pre>
</div>
</div>
<div class="paragraph">
<p>We see that there are ~53k paths in the graph with a similar step count to our input graph, so the paths are much more fragmented.
The compressed mode heatmap is very similar to our full graph.</p>
</div>
<div class="imageblock">
<div class="content">
<img src="../../../assets/mhc_odgi_viz.privvg_10k_100x.compr.png" alt="odgi viz compressed of privvg MHC 10kbp">
</div>
</div>
<div class="paragraph">
<p>Let&#8217;s try to get longer haplotypes, say 50kbp.</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="highlight"><code class="language-shell" data-lang="shell">odgi priv -i mhc.og -o mhc_priv.og -t 16 -b 50000 -d 100 -e 0.01 -P
odgi viz -i mhc_priv.og -o x.png -O</code></pre>
</div>
</div>
<div class="imageblock">
<div class="content">
<img src="../../../assets/mhc_odgi_viz.privvg_50k_100x.compr.png" alt="odgi viz compressed of privvg MHC 50kbp">
</div>
</div>
<div class="paragraph">
<p>And 100kbp:</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="highlight"><code class="language-shell" data-lang="shell">odgi priv -i mhc.og -o mhc_priv.og -t 16 -b 100000 -d 100 -e 0.01 -P
odgi viz -i mhc_priv.og -o x.png -O</code></pre>
</div>
</div>
<div class="imageblock">
<div class="content">
<img src="../../../assets/mhc_odgi_viz.privvg_100k_100x.compr.png" alt="odgi viz compressed of privvg MHC 100kbp">
</div>
</div>
<div class="paragraph">
<p>Huge holes have shown up in the view!
This means that there are parts of the graph which we were not able to sample.
The cause is straightforward: these regions do not contain any exactly shared haplotypes of 100kbp.
Thus, our haplotype frequency filter <code>odgi priv -c 2</code> is preventing us from seeing any of them.
This indicates that for us to obtain a differentially private synthetic pangenome with very long haplotypes across the MHC, we will need to have many more individuals in the input pangenome.
With more individuals, we would expect a higher chance that a <em>group</em> of them has the exact same haplotype across one of these dropout loci.</p>
</div>
<div class="paragraph">
<p>Of course, haplotypes of even 10kbp are very useful.
And for practical purposes, even 1kbp is a reasonable length for many downstream analyses.
The practical utility of a differentially private pangenome is high in the MHC, so long as we can sample shorter haplotypes.
The MHC represents one of the most extreme cases in the pangenome in terms of diversity and low rates of haplotype identity.
This suggests that it should be straightforward to build a differentially private pangenome from the draft human pangenome!</p>
</div>
</div>
<div class="sect2">
<h3 id="_what_about_epsilon">What about \(\epsilon\)?</h3>
<div class="paragraph">
<p>One parameter defines how noisy our transformation is: \(\epsilon\).
A low \(\epsilon\) is likely to be private, but we may introduce too much noise for the database to be useful.
So we want to consider changing this parameter and looking at its effect on the output.</p>
</div>
<div class="paragraph">
<p>When \(\epsilon\) is small, it means that it is difficult for an adversary to distinguish, for pairs of adjacent databases, the database which includes or does not include an individual.
In other words, \(\epsilon\) defines the degree of noise, noisy sampling, or blurring that is added to differentially private queries or synthetic databases.
When the amount of noise is large (when \(\epsilon\) is small), it overwhelms the effects of addition or subtraction of a single individual, thus making it impossible for the adversary to distinguish individuals in general <a href="#dwork2014">[dwork2014]</a>.</p>
</div>
<div class="paragraph">
<p>Let&#8217;s sweep \(\epsilon\) and look at a population genetic parameter of the output graph, the allele frequency.
In our graphs, allele frequency is equivalent to path depth over nodes (which in the pangenome variation graph are sequences, which are equivalent to alleles).
We can extract it with <code>odgi depth -d</code> and plot it with <a href="https://www.r-project.org/"><code>R</code></a>.</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="highlight"><code class="language-shell" data-lang="shell">for e in 0.001 0.01 0.1 1;
do
    odgi priv -i mhc.og -o - -t 16 -b 10000 -d 100 -e $e -P \
      | odgi depth -i - -d &gt;mhc.e$e.tsv &amp;&amp; \
    Rscript -e 'x &lt;- read.delim("'mhc.e$e.tsv'"); require(tidyverse); ggplot(subset(x,depth.uniq&gt;0), aes(x=depth.uniq)) + geom_histogram(binwidth=1) + xlim(0,200); ggsave("'mhc.e$e.tsv.hist.png'", height=3, width=9)'
done</code></pre>
</div>
</div>
<div class="paragraph">
<p>The input allele/node frequency spectrum shows a sharp peak at ~90, which is the number of haplotypes in the pangenome.</p>
</div>
<div class="imageblock">
<div class="content">
<img src="../../../assets/mhc.tsv.hist.png" alt="MHC graph allele frequency spectrum">
</div>
</div>
<div class="paragraph">
<p>But we do not see a similar pattern in a 100x-sampled privvg when using a small \(\epsilon = 0.001\).
We actually see a rather flat frequency distribution, which hides the peaks seen in the real data:</p>
</div>
<div class="imageblock">
<div class="content">
<img src="../../../assets/mhc.e0.001.tsv.hist.png" alt="MHC privvg e=0.001 allele frequency spectrum">
</div>
</div>
<div class="paragraph">
<p>The pattern is similar for \(\epsilon = 0.01\):</p>
</div>
<div class="imageblock">
<div class="content">
<img src="../../../assets/mhc.e0.01.tsv.hist.png" alt="MHC privvg e=0.01 allele frequency spectrum">
</div>
</div>
<div class="paragraph">
<p>At \(\epsilon = 0.1\), we start to observe more low-frequency alleles, consistent with a more relaxed privacy guarantee:</p>
</div>
<div class="imageblock">
<div class="content">
<img src="../../../assets/mhc.e0.1.tsv.hist.png" alt="MHC privvg e=0.1 allele frequency spectrum">
</div>
</div>
<div class="paragraph">
<p>And with \(\epsilon = 1\), we begin to see a peak around the target depth (100x):</p>
</div>
<div class="imageblock">
<div class="content">
<img src="../../../assets/mhc.e1.tsv.hist.png" alt="MHC privvg e=1 allele frequency spectrum">
</div>
</div>
<div class="paragraph">
<p>This becomes progressively more prononced as we increase \(\epsilon = 10\):</p>
</div>
<div class="imageblock">
<div class="content">
<img src="../../../assets/mhc.e10.tsv.hist.png" alt="MHC privvg e=10 allele frequency spectrum">
</div>
</div>
<div class="paragraph">
<p>And for \(\epsilon = 100\), the pattern is starting to be reminiscient of the kind of distribution we saw in the input data, with the peak near 100x rather than 90x of course:</p>
</div>
<div class="imageblock">
<div class="content">
<img src="../../../assets/mhc.e100.tsv.hist.png" alt="MHC privvg e=100 allele frequency spectrum">
</div>
</div>
</div>
<div class="sect2">
<h3 id="_the_effect_of_changing_epsilon">The effect of changing \(\epsilon\)</h3>
<div class="paragraph">
<p>By increasing \(\epsilon\) we are exponentially down-weighting rare haplotypes.
We don&#8217;t follow them because the weight of a given option, \(\exp(\epsilon u(x, r) / \Delta u)\), becomes exponentially larger with higher utility (a.k.a. <code>log1p(haplotype_frequency)</code>).
In effect, we often run down the graph sampling the major allele at every site.
The sampled haplotypes can become very long and uniform in length, because they tend to always match our minimum frequency threshold requirements and run to our target length.</p>
</div>
<div class="paragraph">
<p>At the same time, when increasing \(\epsilon\), we observe an increase in rare alleles in the privvg synthetic pangenome.
This pattern is consistent with a reduction in differential privacy guarantees.
Rare alleles support the identification of individuals.</p>
</div>
<div class="paragraph">
<p>These patterns gives the impression that \(\epsilon\) is behaving as expected, although it&#8217;s worth noting that the interaction between our utility function and privacy is also somewhat confusing.
We will already tend to over-weight common haplotypes due to them having greater utility by our definition.</p>
</div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_consclusions_and_next_steps">Consclusions and next steps</h2>
<div class="sectionbody">
<div class="paragraph">
<p>We&#8217;ve provided a practical first implementation of differential privacy on pangenome graphs.
Our approach is simple and largely follows standard concepts from the differential privacy literature, but the unique application context has made its development require a substantial amount of study.</p>
</div>
<div class="paragraph">
<p>Although we have demonstrated that our implementation in <code>odgi priv</code> follows expected behavior with respect to several key parameters, further evaluation will be required to ascertain the degree to which it provides the desired differential privacy behavior.</p>
</div>
<div class="sect2">
<h3 id="_final_deliverables">Final deliverables</h3>
<div class="paragraph">
<p>We plan to close the project with two final deliverables that build on our implementation.</p>
</div>
<div class="paragraph">
<p>First, we will develop a privvg model from the HPRCy1 draft pangenome.
We will apply methods like the <a href="https://github.com/refresh-bio/agc">assembled genomes compressor <code>agc</code></a> to measure the effect of different parameters on the whole pangenome behavior of privvg.
Because the HPRC is fully public, the utility of this resource is merely research-related, but it will be the first attempt to build such a model for a whole collection of completely assembled genomes.</p>
</div>
<div class="paragraph">
<p>Second, we will establish a set of synthetic "pangenome pairs", where we add and remove specific individuals.
We will present these as a public competition, with a nominal prize for anyone who is able to determine which individuals have been removed from which pangenome model.
By soliciting empirically-driven responses and workflows that describe the privacy-adversarial attacks, we will build practice around the problem of differential privacy on pangenomes.
In effect, this will attach a bounty on the quality of our approach, initating a long-running audit that will establish the practical foundation for applying privvg to real cohort data.</p>
</div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_funding">Funding</h2>
<div class="sectionbody">
<div class="paragraph">
<p><a href="https://nlnet.nl/project/VariationGraph/">This project has been funded</a> through the <a href="https://nlnet.nl/discovery">NGI0 Discovery Fund</a>, a fund established by <a href="https://nlnet.nl/">NLnet</a> with financial support from the European Commission&#8217;s <a href="https://ngi.eu/">Next Generation Internet</a> programme, under the aegis of DG Communications Networks, Content and Technology under grant agreement No 825322.</p>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_references">References</h2>
<div class="sectionbody">
<div class="ulist bibliography">
<ul class="bibliography">
<li>
<p><a id="dwork2014"></a>[dwork2014] Dwork, Cynthia, and Aaron Roth. 2014. “The Algorithmic Foundations of Differential Privacy.” Foundations and Trends in Theoretical Computer Science 9 (3–4): 211–407. <a href="https://privacytools.seas.harvard.edu/files/privacytools/files/the_algorithmic_foundations_of_differential_privacy_0.pdf" class="bare">https://privacytools.seas.harvard.edu/files/privacytools/files/the_algorithmic_foundations_of_differential_privacy_0.pdf</a></p>
</li>
<li>
<p><a id="guarracino2022"></a>[guarracino2022] Guarracino, Andrea, Simon Heumos, Sven Nahnsen, Pjotr Prins, and Erik Garrison. 2022. “ODGI: Understanding Pangenome Graphs.” Bioinformatics , May. <a href="https://doi.org/10.1093/bioinformatics/btac308" class="bare">https://doi.org/10.1093/bioinformatics/btac308</a>.</p>
</li>
<li>
<p><a id="eizenga2020"></a>[eizenga2020] Eizenga, Jordan M., Adam M. Novak, Jonas A. Sibbesen, Simon Heumos, Ali Ghaffaari, Glenn Hickey, Xian Chang, et al. 2020. “Pangenome Graphs.” Annual Review of Genomics and Human Genetics 21 (August): 139–62. <a href="http://dx.doi.org/10.1146/annurev-genom-120219-080406" class="bare">http://dx.doi.org/10.1146/annurev-genom-120219-080406</a></p>
</li>
<li>
<p><a id="liao2022"></a>[liao2022] Liao, Wen-Wei, Mobin Asri, Jana Ebler, Daniel Doerr, Marina Haukness, Glenn Hickey, Shuangjia Lu, et al. 2022. “A Draft Human Pangenome Reference.” bioRxiv. <a href="https://doi.org/10.1101/2022.07.09.499321" class="bare">https://doi.org/10.1101/2022.07.09.499321</a>.</p>
</li>
<li>
<p><a id="garrison2022"></a>[garrison2022] Garrison, Erik, and Andrea Guarracino. 2022. “Unbiased Pangenome Graphs.” bioRxiv. <a href="https://doi.org/10.1101/2022.02.14.480413" class="bare">https://doi.org/10.1101/2022.02.14.480413</a>.</p>
</li>
</ul>
</div>
</div>
</div>
</div>
<div id="footer">
<div id="footer-text">
Last updated 2022-09-24 15:57:58 +0200
</div>
</div>
<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  messageStyle: "none",
  tex2jax: {
    inlineMath: [["\\(", "\\)"]],
    displayMath: [["\\[", "\\]"]],
    ignoreClass: "nostem|nolatexmath"
  },
  asciimath2jax: {
    delimiters: [["\\$", "\\$"]],
    ignoreClass: "nostem|noasciimath"
  },
  TeX: { equationNumbers: { autoNumber: "none" } }
})
MathJax.Hub.Register.StartupHook("AsciiMath Jax Ready", function () {
  MathJax.InputJax.AsciiMath.postfilterHooks.Add(function (data, node) {
    if ((node = data.script.parentNode) && (node = node.parentNode) && node.classList.contains("stemblock")) {
      data.math.root.display = "block"
    }
    return data
  })
})
</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/MathJax.js?config=TeX-MML-AM_HTMLorMML"></script>
</body>
</html>